src/libos/crates/untrusted_allocator/src/allocator/mod.rs

```diff
@@ -0,0 +1,100 @@
+
+// The address of a block returned by malloc or realloc in GNU systems is always a multiple of eight (or sixteen on 64-bit systems).
+const DEFAULT_ALIGNMENT: usize = 16;
+
+use vm_range::*;
+use vm_util::*;
+
+use libc::c_void;
+use libc::{MAP_ANONYMOUS, MAP_PRIVATE, PROT_READ, PROT_WRITE};
+use spin::Mutex;
+use vm_chunk_manager::ChunkManager as VMManager;
+cfg_if::cfg_if! {
+    if #[cfg(feature = "sgx")] {
+        use libc::ocall::{mmap, munmap};
+        use sgx_trts::trts::rsgx_raw_is_outside_enclave;
+    } else {
+        use libc::{mmap, munmap};
+    }
+}
+
+pub struct Allocator {
+    range: VMRange,
+    inner: Mutex<VMManager>,
+}
+
+impl Allocator {
+    pub unsafe fn new(size: usize) -> Self {
+        let total_bytes = size;
+        let start_address = {
+            let addr = unsafe {
+                mmap(
+                    0 as *mut _,
+                    total_bytes,
+                    PROT_READ | PROT_WRITE,
+                    MAP_PRIVATE | MAP_ANONYMOUS,
+                    0,
+                    0,
+                ) as usize
+            };
+
+            debug_assert!(addr.checked_add(total_bytes).is_some());
+            cfg_if::cfg_if! {
+                if #[cfg(feature = "sgx")] {
+                    debug_assert!(rsgx_raw_is_outside_enclave(addr as *const u8, total_bytes));
+                }
+            }
+            addr
+        };
+        let range = VMRange::new(start_address, start_address + total_bytes)
+            .expect("Creating untrusted allocator instance failure");
+        let inner = Mutex::new(
+            VMManager::from(range.start(), total_bytes).expect("creating inner structure failure"),
+        );
+        debug!(
+            "[untrusted alloc] Initiate a new allocator, range = {:?}",
+            range
+        );
+        Self { range, inner }
+    }
+
+    // Allocate a block of memory and return the start address in c style.
+    // Use exactly like malloc from libc.
+    pub unsafe fn alloc(&self, size: usize, align: Option<usize>) -> Result<*mut c_void> {
+        if size > self.range.size() {
+            return_errno!(ENOMEM, "malloc size too big");
+        }
+
+        let align = align.unwrap_or(DEFAULT_ALIGNMENT);
+        let start_addr = self.inner.lock().alloc(size, align)?;
+        Ok(start_addr as *mut c_void)
+    }
+
+    // Free the memory block with specified start address
+    // For memory allocated with malloc_raw, free must be called.
+    // Use exactly like free from libc.
+    pub unsafe fn free(&self, addr: usize) {
+        self.inner.lock().free(addr).expect("free failure");
+    }
+}
+
+impl Drop for Allocator {
+    fn drop(&mut self) {
+        debug!("[untrusted alloc] Drop allocator");
+        assert!(self.inner.lock().check_empty() == true);
+        assert!(self.inner.lock().is_free_range(&self.range));
+        let ret = unsafe { munmap(self.range.start as *mut c_void, self.range.size()) };
+        debug_assert!(ret == 0);
```

Comment:
Do you have a clear rules in mind that instruct you when to use `debug_assert` and when to use `assert`? The codebase is mixed with `debug_assert` and `assert` (for example, this function).

I recommend the following practice.
* For most defensive programming, which are intended to capture logic errors, use `debug_assert`. This way, we can add as many assertions as we want without harming the performance of release builds.
* For defense against Iago attack, always use `assert`.
* For other cases, make judgement on a case-by-case basis.

When following this guidance, most assertion statements would use `debug_assert`. By the way, this is the practice followed in std.

Reply:
Got it. Thank you for summerize the rules. Done.